src/cargo/sources/git/utils.rs

```diff
@@ -344,18 +344,43 @@ impl<'a> GitCheckout<'a> {
     }
 
     fn update_submodules(&self, cargo_config: &Config) -> CargoResult<()> {
-        return update_submodules(&self.repo, cargo_config);
+        // `location` looks like `target/git/checkouts/crate-name-checksum/shorthash`
+        // `submodule_root` looks like `target/git/checkouts/submodules`
+        let checkout_root = self.location.parent().unwrap().parent().unwrap();
+        // Share the same submodules between all checkouts. Without a shared path,
+        // cargo would reclone the submodule for each commit that's checked out,
+        // even if the submodule itself hasn't changed.
+        let submodule_root = checkout_root.join("submodules");
+
+        return update_submodules(&self.repo, cargo_config, &submodule_root);
+
+        fn link_dir(original: &Path, link: &Path) -> std::io::Result<()> {
+            #[cfg(unix)]
+            {
+                std::os::unix::fs::symlink(original, link)
+            }
+            #[cfg(windows)]
+            {
+                std::os::windows::fs::symlink_dir(original, link)
+            }
+        }
```

Comment:
Someone mentioned that this probably needs a fallback for other OSs. `std::fs::soft_link` uses this implementation: ```rust
fs_imp::symlink(original.as_ref(), link.as_ref())
```
and `fs_imp` is doing some complicated things with cfg_if:
```rust
cfg_if::cfg_if! {
    if #[cfg(unix)] {
        mod unix;
        pub use self::unix::*;
    } else if #[cfg(windows)] {
        mod windows;
        pub use self::windows::*;
    } else if #[cfg(target_os = "solid_asp3")] {
        mod solid;
        pub use self::solid::*;
    } else if #[cfg(target_os = "hermit")] {
        mod hermit;
        pub use self::hermit::*;
    } else if #[cfg(target_os = "wasi")] {
        mod wasi;
        pub use self::wasi::*;
    } else if #[cfg(target_family = "wasm")] {
        mod wasm;
        pub use self::wasm::*;
    } else if #[cfg(all(target_vendor = "fortanix", target_env = "sgx"))] {
        mod sgx;
        pub use self::sgx::*;
    } else {
        mod unsupported;
        pub use self::unsupported::*;
    }
}
```
I wonder if it makes sense to *only* special-case `std::os::windows::symlink_dir` and use `soft_link` for the other platforms to avoid having to duplicate the code?